apps/token-manager/test/tokenmanager.js

```diff
@@ -31,17 +33,125 @@ contract('Token Manager', accounts => {
+            const executionTarget = await ExecutionTarget.new()
+            await tokenManager.mint(holder, 100)
+
+            const action = { to: executionTarget.address, calldata: executionTarget.contract.execute.getData() }
+            const script = encodeScript([action])
+
+            await tokenManager.forward(script, { from: holder })
+            assert.equal(await executionTarget.counter(), 1, 'should have received execution call')
+
+            return assertRevert(async () => {
+                await tokenManager.forward(script, { from: accounts[8] })
+            })
+        })
+    })
+
+    context('holder logging', async () => {
+        beforeEach(async () => {
+            await token.changeController(tokenManager.address)
+            await tokenManager.initialize(token.address, true, 0, true)
+        })
+
+        it('logs token manager on issue', async () => {
+            await tokenManager.issue(10)
+
+            const holders = await tokenManager.allHolders()
+            assert.deepEqual(holders, [tokenManager.address], 'holder list should be correct')
+            assert.equal(await tokenManager.holders(0), tokenManager.address, 'should be first holder')
+        })
+
+        it('logs on mints and transfers', async () => {
+            await tokenManager.mint(holder, 10)
+            await token.transfer(accounts[8], 5, { from: holder })
+            await token.transfer(accounts[9], 5, { from: accounts[8] })
+
+            const holders = await tokenManager.allHolders()
+            assert.deepEqual(holders, [holder, accounts[8], accounts[9]], 'holder list should be correct')
+        })
+    })
+
+    context('maximum tokens per address limit', async () => {
+        const limit = 100
 
         beforeEach(async () => {
             await token.changeController(tokenManager.address)
-            await tokenManager.initializeNative(token.address)
+            await tokenManager.initialize(token.address, true, limit, false)
+        })
+
+        it('can mint up to than limit', async () => {
+            await tokenManager.mint(holder, limit)
+
+            assert.equal(await token.balanceOf(holder), limit, 'should have tokens')
+        })
+
+        it('fails to mint more than limit', async () => {
+            return assertRevert(async () => {
+                await tokenManager.mint(holder, limit + 1)
+            })
+        })
+
+        it('can issue unlimited tokens for manager', async () => {
+            await tokenManager.issue(limit + 100000)
+
+            assert.equal(await token.balanceOf(tokenManager.address), limit + 100000, 'should have tokens')
+        })
+
+        it('can assign up to limit', async () => {
+            await tokenManager.issue(limit)
+            await tokenManager.assign(holder, limit)
+
+            assert.equal(await token.balanceOf(holder), limit, 'should have tokens')
+        })
+
+        it('cannot assing more than limit', async () => {
```

Comment:
Actually kind of hilarious, can we `:%s/assign/assing/g`? 😲

Reply:
ENGLISH IS HARD ;)